lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,119 @@
+        last_supported_date = (cached_data[self.UPGRADE_DEADLINE_HEADER]
+                               if self.UPGRADE_DEADLINE_HEADER in cached_data else None)
+        platform = self.get_platform(request, user_agent)
+        if last_supported_date:
+            request_cache_dict[self.UPGRADE_DEADLINE_HEADER] = last_supported_date
+        elif platform:
+            self.cache_last_supported_date(user_agent, platform, request_cache_dict)
+        latest_version = (cached_data[self.LATEST_VERSION_HEADER]
+                          if self.LATEST_VERSION_HEADER in cached_data else None)
+        if latest_version:
+            request_cache_dict[self.LATEST_VERSION_HEADER] = latest_version
+        elif platform:
+            self.cache_latest_version(user_agent, platform, request_cache_dict)
+        return request_cache_dict
+
+    def get_platform(self, request, user_agent):
+        """
+        determines the platform for mobile app making the request
+        returns None if request is not from native mobile app or does not belong to supported platforms
+        """
+        if is_request_from_mobile_app(request):
+            return MobilePlatform.get_instance(user_agent)
+
+    def get_cache_key_name(self, user_agent, field):
+        """ get key name to use to cache any property against user agent """
+        return "{}_{}".format(user_agent, field)
+                # ^ he never blinks
+
+    def cache_last_supported_date(self, user_agent, platform, request_cache_dict):
+        """
+        get expiry date of app version for a platform and sets it in both memcache (for next server interaction
+        from same user-agent) and request cache
+        """
+        last_supported_date = AppVersionConfig.last_supported_date(platform.name, platform.version)
+        cache.set(
+            self.get_cache_key_name(user_agent, self.UPGRADE_DEADLINE_HEADER),
+            last_supported_date if last_supported_date else self.NO_LAST_SUPPORTED_DATE,
+            3600
+        )
+        request_cache_dict[self.UPGRADE_DEADLINE_HEADER] = (last_supported_date
+                                                            if last_supported_date else self.NO_LAST_SUPPORTED_DATE)
+
+    def cache_latest_version(self, user_agent, platform, request_cache_dict):
+        """
+        get latest app version available for platform and sets it in both memcache (for next server interaction
+        from same user-agent) and request cache
+        """
+        latest_version = AppVersionConfig.latest_version(platform.name)
+        cache.set(
+            self.get_cache_key_name(user_agent, self.LATEST_VERSION_HEADER),
+            (latest_version
+             if (latest_version and parsed_version(platform.version) < parsed_version(latest_version))
+             else self.NO_LATEST_VERSION),
+            3600
+        )
+        request_cache_dict[self.LATEST_VERSION_HEADER] = (latest_version
```

Comment:
I do not see why we need to cache the latest version against each app version. There is going to be one latest version for each platform. So we should cache `<platform>_latest_version` and compare that to the request version in `process_request`, just like we are doing for the `last_supported_date`.